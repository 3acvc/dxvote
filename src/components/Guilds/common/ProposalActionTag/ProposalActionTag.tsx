import React from 'react';
import styled, { css } from 'styled-components';

export const ActionTypes = {
  for: 'for',
  against: 'against',
};

const getActionLabel = type => {
  return {
    [ActionTypes.for]: 'For',
    [ActionTypes.against]: 'Against',
  }[type];
};

interface ProposalActionTagProps {
  type: string | 'for' | 'against';
  label?: string;
}
const variantStyles = type =>
  ({
    [ActionTypes.for]: css`
      color: ${({ theme }) => theme.blue};
      border-color: ${({ theme }) => theme.blue};
    `,
    [ActionTypes.against]: css`
      color: ${({ theme }) => theme.red};
      border-color: ${({ theme }) => theme.red};
    `,
  }[type]);

const Tag = styled.span<ProposalActionTagProps>`
  margin: 0;
  border-radius: 6px;
  padding: 4px 8px;
  border: 1px solid ${({ theme }) => theme.colors.muted};
  font-style: normal;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  ${({ type }) => variantStyles(type)}
`;

export const ProposalActionTag: React.FC<ProposalActionTagProps> = ({
  type,
  label,
}) => {
  if (!Object.values(ActionTypes).includes(type)) return null;
  return <Tag type={type}>{label ? label : getActionLabel(type)}</Tag>;
};
