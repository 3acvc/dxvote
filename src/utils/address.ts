import { ethers, utils } from 'ethers';
const appConfig = require('../config.json');

export function shortenAddress(address, digits = 4) {
  if (!isAddress(address)) {
    throw Error(`Invalid 'address' parameter '${address}'.`);
  }
  return `${address.substring(0, digits + 2)}...${address.substring(
    42 - digits
  )}`;
}

export function isAddress(value) {
  try {
    return ethers.utils.getAddress(value.toLowerCase());
  } catch {
    return false;
  }
}

export function toChecksum(address) {
  return utils.getAddress(address);
}

export function toAddressStub(address, size = 'default') {
  const start = address.slice(0, 6);
  const end = address.slice(-4);

  switch (size) {
    case 'short':
      return `${start}..`;
    case 'long':
      return address;
    default:
      return `${start}...${end}`;
  }
}

export function getBlockchainLink(address, networkName, type) {
  switch (type) {
    case 'user':
      return `${window.location.pathname}#/user/${address}`;
    case 'address':
      if (networkName === 'arbitrum')
        return `https://explorer.arbitrum.io/#/address/${address}`;
      else if (networkName === 'arbitrumTestnet')
        return `https://rinkeby-explorer.arbitrum.io/#/address/${address}`;
      else if (networkName === 'mainnet')
        return `https://etherscan.io/address/${address}`;
      else if (networkName === 'xdai')
        return `https://blockscout.com/xdai/mainnet/address/${address}`;
      else return `https://${networkName}.etherscan.io/address/${address}`;
    default: // investigate DRY here
      if (networkName === 'arbitrum')
        return `https://explorer.arbitrum.io/#/tx/${address}`;
      else if (networkName === 'arbitrumTestnet')
        return `https://rinkeby-explorer.arbitrum.io/#/tx/${address}`;
      else if (networkName === 'mainnet')
        return `https://etherscan.io/tx/${address}`;
      else if (networkName === 'xdai')
        return `https://blockscout.com/xdai/mainnet/tx/${address}`;
      else return `https://${networkName}.etherscan.io/tx/${address}`;
  }
}

export async function getENSName(address) {
  let name = null;
  try {
    const provider = ethers.getDefaultProvider();
    const checksumed = ethers.utils.getAddress(address);
    name = provider.lookupAddress(checksumed);
  }
  catch (e) {
    console.error(e);
  }
  return name;
} 

export function getERC20Token(address) {
  let tokenObject;
  Object.keys(appConfig).forEach((network => {
    tokenObject = appConfig[network]?.tokens?.filter(token => token.address === address)
    if (tokenObject) return;
  }));
  return tokenObject;
}

export function getDxVoteContract(address) {
  let obj = null;
  Object.keys(appConfig).forEach(network => {
    let contracts = appConfig[network]?.contracts
    for (let [key, value] of Object.entries(contracts)) {
      if (value === address) {
        obj = {contract: key, address: value}
        return;
      }
    }
    return;
  });
  return obj;
}