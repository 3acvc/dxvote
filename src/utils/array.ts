export const flatten = (arr: any[]) =>
  arr.reduce(
    (acc, val) => acc.concat(Array.isArray(val) ? flatten(val) : val),
    []
  );
