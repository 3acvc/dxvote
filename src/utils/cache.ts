import contentHash from 'content-hash';
import _ from 'lodash';

const Web3 = require('web3');
const web3 = new Web3();

const MAX_BLOCKS_PER_EVENTS_FETCH: number =
  Number(process.env.MAX_BLOCKS_PER_EVENTS_FETCH) || 100000;

export const getEvents = async function (
  web3,
  contract,
  fromBlock,
  toBlock,
  eventsToGet,
  maxBlocksPerFetch = MAX_BLOCKS_PER_EVENTS_FETCH
) {
  let events = [],
    to = Math.min(fromBlock + maxBlocksPerFetch, toBlock),
    from = fromBlock;
  while (from < to) {
    console.debug(
      `Fetching events of ${contract._address} from blocks ${from} -> ${to}`
    );
    try {
      let eventsFetched = await contract.getPastEvents(eventsToGet, {
        fromBlock: from,
        toBlock: to,
      });
      eventsFetched = await getTimestampOfEvents(web3, eventsFetched);
      events = events.concat(eventsFetched);
      from = to;
      to = Math.min(from + maxBlocksPerFetch, toBlock);
    } catch (error) {
      console.error('Error fetching blocks:', (error as Error).message);
      if (Math.trunc((to - from) / 2) > 10000) {
        const blocksToLower = Math.max(Math.trunc((to - from) / 2), 10000);
        console.debug('Lowering toBlock', blocksToLower, 'blocks');
        to = to - blocksToLower;
      }
    }
  }
  return events;
};

export const getRawEvents = async function (
  web3,
  contractAddress,
  fromBlock,
  toBlock,
  topicsToGet,
  maxBlocksPerFetch = MAX_BLOCKS_PER_EVENTS_FETCH
) {
  let events = [],
    to = Math.min(fromBlock + maxBlocksPerFetch, toBlock),
    from = fromBlock;
  while (from < to) {
    console.debug(
      `Fetching logs of ${contractAddress} from blocks ${from} -> ${to}`
    );
    try {
      let eventsFetched = await web3.eth.getPastLogs({
        address: contractAddress,
        fromBlock: from,
        toBlock: to,
        topics: topicsToGet,
      });
      eventsFetched = await getTimestampOfEvents(web3, eventsFetched);
      events = events.concat(eventsFetched);
      from = to;
      to = Math.min(from + maxBlocksPerFetch, toBlock);
    } catch (error) {
      console.error('Error fetching blocks:', (error as Error).message);
      if (Math.trunc((to - from) / 2) > 10000) {
        const blocksToLower = Math.max(Math.trunc((to - from) / 2), 10000);
        console.debug('Lowering toBlock', blocksToLower, 'blocks');
        to = to - blocksToLower;
      }
    }
  }
  return events;
};

export const getTimestampOfEvents = async function (web3, events) {
  //// TODO:  See how can we bacth requests can be implemented

  // async function batchRequest(blocks) {
  //   const batch = new web3.BatchRequest();
  //   let requests = [];
  //   for (let i = 0; i < blocks.length; i++) {
  //     const request = new Promise((resolve, reject) => {
  //       batch.add(web3.eth.getBlock.request(blocks[i], (err, data) => {
  //         console.log(1)
  //         if (err) return reject(err);
  //         resolve(data);
  //       }));
  //     });
  //     requests.push(request);
  //   }
  //   batch.execute();
  //   console.log(batch)
  //   await Promise.all(requests);
  //   return batch;
  // };

  let blocksToFetch = [];
  let timestamps = [];
  events.map(event => {
    if (blocksToFetch.indexOf(event.blockNumber) < 0)
      blocksToFetch.push(event.blockNumber);
  });
  const totalLength = blocksToFetch.length;
  while (blocksToFetch.length > 0 && totalLength > timestamps.length) {
    // timestamps = (await batchRequest(blocksToFetch)).map((blockResult) => {
    //   return blockResult.timestamp;
    // });
    const blocksToFetchBatch = blocksToFetch.splice(0, 500);
    await Promise.all(
      blocksToFetchBatch.map(async block => {
        const blockInfo = await web3.eth.getBlock(block);
        for (let i = 0; i < events.length; i++) {
          if (events[i].blockNumber === blockInfo.number)
            events[i].timestamp = blockInfo.timestamp;
          if (blockInfo.l2BlockNumber)
            events[i].blockNumber = Number(blockInfo.l2BlockNumber);
        }
      })
    );
  }
  return events;
};

export const sortEvents = function (events) {
  return _.orderBy(
    events,
    ['blockNumber', 'transactionIndex', 'logIndex'],
    ['asc', 'asc', 'asc', 'asc']
  );
};

export const executeMulticall = async function (web3, multicall, calls) {
  const rawCalls = calls.map(call => {
    return [
      call[0]._address,
      web3.eth.abi.encodeFunctionCall(
        call[0]._jsonInterface.find(method => method.name === call[1]),
        call[2]
      ),
    ];
  });

  const { returnData } = await multicall.methods.aggregate(rawCalls).call();

  return {
    returnData,
    decodedReturnData: returnData.map((callResult, i) => {
      return web3.eth.abi.decodeParameters(
        calls[i][0]._jsonInterface.find(method => method.name === calls[i][1])
          .outputs,
        callResult
      )['0'];
    }),
  };
};

export const executeRawMulticall = async function (multicall, calls) {
  const rawCalls = calls.map(call => {
    return [call[0], call[1]];
  });

  const { returnData } = await multicall.methods.aggregate(rawCalls).call();

  return {
    returnData,
  };
};

export const isNode = function () {
  return typeof module !== 'undefined' && module.exports;
};

export const descriptionHashToIPFSHash = function (descriptionHash) {
  try {
    if (contentHash.getCodec(descriptionHash) === 'ipfs-ns')
      return contentHash.decode(descriptionHash);
    else if (
      descriptionHash.length > 1 &&
      descriptionHash.substring(0, 2) !== 'Qm'
    )
      return descriptionHash;
    else return '';
  } catch (error) {
    return '';
  }
};

export const ipfsHashToDescriptionHash = function (ipfsHash) {
  try {
    if (ipfsHash.length > 1 && ipfsHash.substring(0, 2) === 'Qm')
      return contentHash.fromIpfs(ipfsHash);
    else if (contentHash.getCodec(ipfsHash) === 'ipfs-ns') return ipfsHash;
    else return '';
  } catch (error) {
    return '';
  }
};

export const getSchemeConfig = function (networkContracts, schemeAddress) {  
  if (networkContracts?.daostack[schemeAddress])
    return {
      type: networkContracts.daostack[schemeAddress].type,
      name: networkContracts.daostack[schemeAddress].name,
      contractToCall: networkContracts.daostack[schemeAddress].contractToCall,
      newProposalTopics: networkContracts.daostack[schemeAddress].newProposalTopics,
      creationLogEncoding: networkContracts.daostack[schemeAddress].creationLogEncoding,
      votingMachine: networkContracts.daostack[schemeAddress].votingMachine,
      voteParams: networkContracts.daostack[schemeAddress].voteParams,
    }
  else 
    return {
      type: 'WalletScheme',
      name: 'WalletScheme',
      contractToCall: schemeAddress,
      newProposalTopics: [
        [
          web3.utils.soliditySha3('ProposalStateChange(bytes32,uint256)'),
          null,
          '0x0000000000000000000000000000000000000000000000000000000000000001',
        ],
      ],
      creationLogEncoding: [],
    };
};

export async function tryCacheUpdates(promises, networkCache) {
  let retry = true;
  while (retry) {
    try {
      (await Promise.all(promises)).map(networkCacheUpdated => {
        networkCache = networkCacheUpdated;
      });
    } catch (e) {
      console.error('[CACHE UPDATE] (trying again)', e.message);
    } finally {
      retry = false;
    }
  }
  return networkCache;
}