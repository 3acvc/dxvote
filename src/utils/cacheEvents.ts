const _ = require("lodash");

const MAX_BLOCKS_PER_EVENTS_FETCH : number = Number(process.env.MAX_BLOCKS_PER_EVENTS_FETCH) || 1000000;

export const getEvents = async function(
  contract, fromBlock, toBlock, eventsToGet, maxBlocksPerFetch = MAX_BLOCKS_PER_EVENTS_FETCH
) {
  let events = [], to = Math.min(fromBlock + maxBlocksPerFetch, toBlock), from = fromBlock;
  while (from < to) {
    console.debug(`Fetching events of ${contract._address} from blocks ${from} -> ${to}`);
    try {
      const eventsFetched = await contract.getPastEvents(eventsToGet, {fromBlock: from, toBlock: to});
      events = events.concat(eventsFetched);
      from = to;
      to = Math.min(from + maxBlocksPerFetch, toBlock);
    } catch (error) {
      console.debug('Lowering toBlock', (to - from) / 2, 'blocks');
      to = from + (to - from) / 2;
    }
  };
  return events;
};

export const getRawEvents = async function(
  web3, contractAddress, fromBlock, toBlock, topicsToGet, maxBlocksPerFetch = MAX_BLOCKS_PER_EVENTS_FETCH
) {
  let events = [], to = Math.min(fromBlock + maxBlocksPerFetch, toBlock), from = fromBlock;
  while (from < to) {
    console.debug(`Fetching logs of ${contractAddress} from blocks ${from} -> ${to}`);
    try {
      const eventsFetched = await web3.eth.getPastLogs({
        address: contractAddress,
        fromBlock: from,
        toBlock: to,
        topics: topicsToGet
      })
      events = events.concat(eventsFetched);
      from = to;
      to = Math.min(from + maxBlocksPerFetch, toBlock);
    } catch (error) {
      console.debug('Lowering toBlock', (to - from) / 2, 'blocks');
      to = from + (to - from) / 2;
    }
  };
  return events;
};

export const sortEvents = function(events) {
  return _.orderBy( events , ["blockNumber", "transactionIndex", "logIndex"], ["asc","asc","asc"]);
};
