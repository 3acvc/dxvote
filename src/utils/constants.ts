export const MAX_UINT =
  '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
export const ZERO_HASH =
  '0x0000000000000000000000000000000000000000000000000000000000000000';
export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
export const ANY_ADDRESS = '0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa';
export const ANY_FUNC_SIGNATURE = '0xaaaaaaaa';
export const ERC20_TRANSFER_SIGNATURE = '0xa9059cbb';
export const ERC20_APPROVE_SIGNATURE = '0x095ea7b3';
export const DEFAULT_TOKEN_DECIMALS = 18;

export const NETWORK_NAMES = {
  '1': 'mainnet',
  '4': 'rinkeby',
  '100': 'xdai',
  '42161': 'arbitrum',
  '421611': 'arbitrumTestnet',
  '1337': 'localhost',
};

export const NETWORK_IDS = {
  mainnet: 1,
  rinkeby: 4,
  xdai: 100,
  arbitrum: 42161,
  arbitrumTestnet: 421611,
  localhost: 1337,
};

export const NETWORK_ASSET_SYMBOL = {
  mainnet: 'ETH',
  rinkeby: 'ETH',
  xdai: 'XDAI',
  arbitrum: 'ETH',
  arbitrumTestnet: 'ETH',
  localhost: 'ETH',
};
