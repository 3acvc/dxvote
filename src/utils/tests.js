import { ThemeProvider } from 'styled-components';
import theme from '../theme/light.json';

export const themedComponent = children => (
  <ThemeProvider theme={theme}> {children} </ThemeProvider>
);
