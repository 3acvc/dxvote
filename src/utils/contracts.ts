import { Contract } from 'ethers';
import { AddressZero } from 'ethers/constants';
import { JsonRpcSigner, Web3Provider } from 'ethers/providers';
import { isAddress } from '.';

function getSigner(library: Web3Provider, account: string): JsonRpcSigner {
  return library.getSigner(account);
}

function getProviderOrSigner(
  library: Web3Provider,
  account?: string
): Web3Provider | JsonRpcSigner {
  return account ? getSigner(library, account) : library;
}

export function getContract(
  address: string,
  ABI: any,
  library: Web3Provider,
  account?: string
): Contract {
  if (!isAddress(address) || address === AddressZero) {
    throw Error(`Invalid 'address' parameter '${address}'.`);
  }

  return new Contract(
    address,
    ABI,
    getProviderOrSigner(library, account) as any
  );
}
