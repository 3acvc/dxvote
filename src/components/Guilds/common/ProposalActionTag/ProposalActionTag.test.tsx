import { render } from '../../../../utils/tests';
import { ProposalActionTag, ActionTypes } from './ProposalActionTag';

describe('ProposalActionTag', () => {
  it('Should match snapshot', () => {
    const { container } = render(<ProposalActionTag type={ActionTypes.for} />);
    expect(container).toMatchSnapshot();
  });
});
