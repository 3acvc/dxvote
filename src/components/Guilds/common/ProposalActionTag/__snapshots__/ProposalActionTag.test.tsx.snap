// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalActionTag Should match snapshot 1`] = `
<div>
   
  <span
    class="sc-bdfBQB kaZTHv"
    type="for"
  >
    For
  </span>
   
</div>
`;
